doc/feature/httpServerREST.md

```diff
@@ -0,0 +1,161 @@
+HTTP Server for generic REST interfaces {#httpserverrest}
+============
+
+This document describes a proposed extension to the current http::Server to make
+it usable as a general-purpose REST interface.
+
+## Requirements
+
+* Support all standard http verbs: GET, POST, PUT, PATCH, DELETE
+* Support custom response payload for all action verbs (not just for GET)
+* Allow specifying the MIME type of the response to serve any type of content
+* Allow specifying extra HTTP header fields in response to certain requests
+* Allow asynchronous processing of requests (not blocking the main thread)
+* Single endpoint for collections (e.g. windows/, files/, neurons/, etc. )
+* Allow filtering operations on collections (windows/selected, files?limit=10)
+
+## Dependency Changes
+
+None, cppnetlib 0.13 already offers the required functionality.
+
+## API
+
+To fullfill the abovementioned requirements with minimal API changes, we propose
+a generic interface as follows:
+
+    namespace zeroeq
+    {
+    namespace http
+    {
+    enum class Verb { GET, POST, PUT, PATCH, DELETE };
+
+    // copied from HTTPServer::connection::status_t
+    enum Code { OK = 200, NOT_FOUND = 404, ... };
+
+    // some commonly used headers, see:
+    // https://en.wikipedia.org/wiki/List_of_HTTP_header_fields
+    enum class Header{ LOCATION, RETRY_AFTER, ... };
+
+    struct Response
+    {
+        Code code = Code::OK;
+        std::string contentType;
+        std::string payload;
+        std::map<Header, std::string> headers;
+    };
+
+    /** HTTP REST callback with payload, returning a Response future. */
+    using RESTFunc = std::function<std::future<Response>(const std::string&)>;
+
+    /** HTTP REST callback with path + payload, returning a Response future. */
+    using RESTPathFunc = std::function<std::future<Response>(const std::string&,
+                                                             const std::string&)>;
```

Comment:
As I understand it with this declarations you are leaving to the client the responsibility of implementing the asynchronous handling of requests.  If a client creates the future with std::async(std::deferred, ...) then the request is not handled asynchronously any more.
Is this deliberate for good reasons? Couldn't we hide the threading in the server internals?
If the server implementation is responsible of calling get in the future, I no good reason to require these function to return a future. The server could wrap in internally in the same way you've done in the first example.